keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+
+* **Guaranteed QoS**
+
+    A pod is considered `Guaranteed`, if for each resource type, either of following is true:
+    * It is explicitly resourced (pod.spec.resources is set) for that resource type AND the pod's request is equal to its limits.
+    * It is implicitly resourced (pod.spec.resources is not set) AND all containers requests are equal to their limits.
+
+  **Note: If the pod spec includes `resources` stanza at pod level, but either requests or limits are missing for a resource, the pod will not be considered Guaranteed QoS, even if container-level requests and limits are present and are equal.**
+
+* **Burstable QoS Criteria**
+
+    A pod is considered `Burstable`, if for each resource type it doesn't meet the criteria for Guaranteed QoS for that resource, and either of following is true:
+
+    * It is explicitly resourced (pod.spec.resources is set) for that resource, AND request (less than or equal to limit) or limit is set.
+    * It is implicitly resourced for that resource AND for at least one container in the pod, request (less than or equal to limit) or limit is set. 
+  
+
+* **BestEffort QoS Criteria**
+
+    None of pod-level or container-level resources requests or limits are specified.
+
+
+#### OOM Killer Behavior
+With Pod level resources, the OOM Killer will evolve to consider the cumulative resource limits set for the entire Pod. If the aggregated
+memory usage of all the containers within a Pod exceeds the effective pod level limits, the entire Pod becomes a candidate for OOM Killing, even if individual containers are within their limits.
+
+Within a pod, the OOM Killer will maintain its existing behavior, considering container level limits (if set) and QoS classes, so a container
+exceeding its individual limits can still be terminated even if a pod as a whole is under its limit.
+
+#### Admission Controller
+
+* LimitRanger
+
+  LimitRanger today supports objects of type `Container`, `Pod` and `PersistentVolumeClaim`. It allows following constraints:
+  
+    * min (this option is to specify min usage constraint on the resource)
+    * max (this option is to specify max usage constraints on the resource)
+    * defaultRequest (this option is to specify default resource request for the named resource)
+    * default (this option is to specify default limit for the named resource)
+    * maxLimitRequestRatio (represents the max burst value of the named resource)
+  
+  default and defaultRequest are not supported by LimitRanger of type `Pod`. To incorporate Pod level resource specifications, LimitRanger of type `Pod` logic will be extended with following changes:
+  1. Validation at Pod Admission will derive pod resources from `resources` stanza at pod level if it is set.
+  2. Extend `Pod` type LimitRanger to support `defaultRequest` and `default` to allow setting Pod level defaults.
```

Comment:
will limit ranger have a consistency check? If container AND Pod defaulters are confiogured, limitranger can make Pod spec be incorrect (it is already can do it in some edge cases, I wonder if we increase the number of those edge cases)

Reply:
Since limit ranger is implemented as a list, I think it's better to keep different limitRangers independent.

But we can look into better practices for beta

Reply:
yeah we can revisit this again for Beta to see if there's a change to reduce the number of edge cases.. But for now, we are complying with what's mentioned in the public doc: "A LimitRange does not check the consistency of the default values it applies." https://kubernetes.io/docs/concepts/policy/limit-range/#limitrange-and-admission-checks-for-pods